crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -823,18 +825,46 @@ impl<'db> TypeInferenceBuilder<'db> {
         self.infer_optional_expression(cause.as_deref());
     }
 
+    /// Given a `from .foo import bar` relative import, resolve the relative module
+    /// we're importing `bar` from into an absolute [`ModuleName`]
+    /// using the name of the module we're currently analyzing.
+    ///
+    /// - `level` is the number of dots at the beginning of the relative module name:
+    ///   - `from .foo import bar` => `level == 1`
+    ///   - `from ...foo import bar` => `level == 3`
+    /// - `tail` is the relative module name stripped of all leading dots:
+    ///   - `from .foo import bar` => `tail == `"foo"`
+    ///   - `from ..foo.bar import baz` => `tail == "foo.bar"`
+    fn relative_module_name(&self, tail: Option<&str>, level: NonZeroU32) -> Option<ModuleName> {
+        let module = file_to_module(self.db, self.file)?;
+        let mut level = u32::from(level);
+        if module.kind().is_package() {
+            level -= 1;
+        }
+        let mut module_name = module.name().to_owned();
+        for _ in 0..level {
+            module_name = module_name.parent()?;
+        }
+        if let Some(tail) = tail {
+            module_name.push_tail(tail).ok()?;
+        }
```

Comment:
I think we should add a TODO here to raise an error if the module name is not a valid identifier or how do we do this for non-relative paths?

Reply:
Other than `from foo import *` imports (which will anyway require their own branch of logic), I don't think it's possible for any components of the module name here to not be a valid identifier. The source code would fail to parse, so we'd never get here. So I'm not sure how to add a test for this.

Ideally the `ruff_python_ast::nodes::Identifier` node would have invariants around it that ensured that it never wrapped a string that wasn't a valid Python identifier: if so, we could also implement a more efficient conversion from `ruff_python_ast::nodes::Identifier` to `ModuleName`, that skips some of the checks performed by the `ModuleName` constructor.

Reply:
Fair, although this makes assumptions about what the parser does. The parser could decide to recover from `a.5.b` because that gives a better overall parse result than stopping after `a.5`. 

I think it would be good to at least log a debug or warning because it would be an unexpected outcome.

Reply:
> Fair, although this makes assumptions about what the parser does. The parser could decide to recover from `a.5.b` because that gives a better overall parse result than stopping after `a.5`.

Okay, but even if it decides to do error recovery, the parser should still mark the node as invalid somehow, correct? And if so, we shouldn't *ever* get here; it would be fruitless to attempt to do type analysis on a module that's known to contain invalid syntax (which is what this would be)

> I think it would be good to at least log a debug or warning because it would be an unexpected outcome.

fair enough

Reply:
> Okay, but even if it decides to do error recovery, the parser should still mark the node as invalid somehow

Ideally yes, and we should handle it here. Adding a log should help us find this if we ever forget.

>  it would be fruitless to attempt to do type analysis on a module that's known to contain invalid syntax (which is what this would be)

I don't agree with this. Doing type inference on programs with invalid syntax is a primary goal and we should do the best we can or Red Knot won't be suited for editors.

Reply:
> I don't agree with this. Doing type inference on programs with invalid syntax is a primary goal and we should do the best we can or Red Knot won't be suited for editors.

Interesting. From our previous discussions I understood that having an error-recoverable parser that could offer helpful error messages in cases of invalid syntax was a primary goal, but that being able to offer diagnostics and perform semantic analysis on modules with invalid syntax was not a goal that we expected to achieve in the foreseeable future. Was my understanding incorrect? Or are you saying that it's only a non-goal for _Ruff_, and that we plan on supporting modules with invalid syntax from the very start with red-knot?

Reply:
We de-prioritized the recoverable semantic analysis work in Ruff, but Red Knot should have support it from the very start. We may want to spend some more time to think about how we want to do that. @carljm general idea is to just return `Unknown` or `Unbound` in these cases but I start to think that it may involve more than just that. For example, we may want to avoid raising exceptions for calls to a function where the function's declaration contains syntax errors because that likely only contributes noise.